### ISSUE TITLE
IEEE754 of size 16 not yet implemented

### ISSUE THREAD

**Issue by pheix** [bbkr/GeoIP2#31]
Hi, Pawel! I noticed yesterday, that MaxMind changed their `GeoLite2-City.mmdb` database and now I get in my [unit tests](https://gitlab.com/pheix/dcms-raku/-/jobs/8730911809):

```
# Failed test 'Check do_log method'
# at ./t/16-stats.t line 106
# Subtest: Check get_country method
    1..27
IEEE754 of size 16 not yet implemented. Sorry.
  in block  at /usr/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 361
  in method read-floating-number at /usr/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 357
  in method read-data at /usr/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 250
  in method read-metadata at /usr/share/perl6/site/precomp/BBDA2339AB6DEDE710FDC2409265B6907452A571/E5/E5A4B16713B62F06C1B07F86768C4F2791F78B68 line 1
  in submethod BUILD at /usr/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 35
  in method get_country at /dcms-raku/lib/Pheix/Controller/Stats.rakumod (Pheix::Controller::Stats) line 65
  in block  at ./t/16-stats.t line 174
  in method get_code_time at /dcms-raku/lib/Pheix/Test/Helpers.rakumod (Pheix::Test::Helpers) line 77
  in block  at ./t/16-stats.t line 169
  in sub subtest at /usr/share/perl6/core/sources/12B9E3A188BF69AFD71E2C4A4AFA59CFF2CEEA91 (Test) line 422
  in block <unit> at ./t/16-stats.t line 136

    # You planned 27 tests, but ran 0
```

I re-built the CI/CD docker yesterday evening, looks like the latest databases were downloaded: https://gitlab.com/pheix-docker/rakudo-star/-/jobs/8728063176#L2500. Just tried it manually in [container](https://gitlab.com/pheix-docker/rakudo-star/container_registry/6729093):

```
root@webtech-omen:/# ls -la /usr/share/GeoIP/
total 62532
drwxr-xr-x 1 root root      112 Dec 26 21:37 .
drwxr-xr-x 1 root root     1150 Dec 26 22:44 ..
-rw------- 1 root root        0 Dec 26 21:37 .geoipupdate.lock
-rw-r--r-- 1 root root 56191986 Dec 26 21:37 GeoLite2-City.mmdb
-rw-r--r-- 1 root root  7838627 Dec 26 21:37 GeoLite2-Country.mmdb
root@webtech-omen:/# raku -MGeoIP2 -e 'GeoIP2.new( path => "/usr/share/GeoIP/GeoLite2-City.mmdb" ).locate(ip => "146.19.120.10")<country><iso_code>.say'
IEEE754 of size 16 not yet implemented. Sorry.
  in block  at /usr/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 361
  in method read-floating-number at /usr/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 357
  in method read-data at /usr/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 250
  in method read-metadata at /usr/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 154
  in submethod BUILD at /usr/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 35
  in block <unit> at -e line 1

root@webtech-omen:/# raku -MGeoIP2 -e 'GeoIP2.new(path => "/usr/share/GeoIP/GeoLite2-Country.mmdb").locate(ip => "146.19.120.10")<country><iso_code>.say'
PL
root@webtech-omen:/#
```

Is it possible to fix it? I will switch to `GeoLite2-Country.mmdb` on my end, but looks like it very possible to get the same error there soon.

Thanks in advance!

**Comment by bbkr** [comment]
Hi. I'm unable to reproduce this error, both databases downloaded directly from MaxMind on `2024-12-27`:

```
$ raku -I lib -e 'use GeoIP2; GeoIP2.new(path => "GeoLite2-Country.mmdb").locate(ip => "146.19.120.10")<country><iso_code>.say';
PL

$ raku -I lib -e 'use GeoIP2; GeoIP2.new(path => "GeoLite2-City.mmdb").locate(ip => "146.19.120.10")<country><iso_code>.say';
PL
```

Do you still have this file? It seems like broken release by MaxMind:
- [The length of a float is always 4 bytes.](https://github.com/maxmind/MaxMind-DB/blob/main/MaxMind-DB-spec.md#float---15)
- [The length of a double is always 8 bytes.](https://github.com/maxmind/MaxMind-DB/blob/main/MaxMind-DB-spec.md#double---3)

So my library should never enter [exception you encountered](https://github.com/bbkr/GeoIP2/blob/master/lib/GeoIP2.rakumod#L361) operating on file matchin their own spec. There are simply no 16 bytes IEEE754 types defined there.

Please try with recent files and let me know if it resolved your issue.

**Comment by pheix** [comment]
Thanks for reply! After the update it works fine:

```
[kostas@webtech-omen ~]$ raku -MGeoIP2 -e 'GeoIP2.new( path => "/usr/share/GeoIP/GeoLite2-City.mmdb" ).locate(ip => "146.19.120.10")<country><iso_code>.say'
IEEE754 of size 16 not yet implemented. Sorry.
  in block  at /opt/rakudo/rakudo-star-2023.05/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 361
  in method read-floating-number at /opt/rakudo/rakudo-star-2023.05/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 357
  in method read-data at /opt/rakudo/rakudo-star-2023.05/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 250
  in method read-metadata at /opt/rakudo/rakudo-star-2023.05/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 154
  in submethod BUILD at /opt/rakudo/rakudo-star-2023.05/share/perl6/site/sources/E5A4B16713B62F06C1B07F86768C4F2791F78B68 (GeoIP2) line 35
  in block <unit> at -e line 1
[kostas@webtech-omen ~]$ geoipupdate -v
geoipupdate version 7.1.0 (662daf83, 2024-11-18T17:41:36Z, linux-amd64)
Using config file /etc/GeoIP.conf
Using database directory /usr/share/GeoIP
Initializing file lock at /usr/share/GeoIP/.geoipupdate.lock
Acquired lock file at /usr/share/GeoIP/.geoipupdate.lock
Calculated MD5 sum for /usr/share/GeoIP/GeoLite2-Country.mmdb: cfbfe1692194d89bccf433687f51c8b0
Updates available for GeoLite2-Country
Database GeoLite2-Country successfully updated: 4ce22566fa6dd97b54fea90f696bcd8d
Calculated MD5 sum for /usr/share/GeoIP/GeoLite2-City.mmdb: 2c33f97afeae3af1635cff9afe3a1074
Updates available for GeoLite2-City
Database GeoLite2-City successfully updated: 8a4f2cdbf8a8ee451a680a5bd2d356b4
Lock file /usr/share/GeoIP/.geoipupdate.lock successfully released
[kostas@webtech-omen ~]$ raku -MGeoIP2 -e 'GeoIP2.new( path => "/usr/share/GeoIP/GeoLite2-City.mmdb" ).locate(ip => "146.19.120.10")<country><iso_code>.say'
PL
[kostas@webtech-omen ~]$ raku -MGeoIP2 -e 'GeoIP2.new( path => "/usr/share/GeoIP/GeoLite2-Country.mmdb" ).locate(ip => "146.19.120.10")<country><iso_code>.say'
PL
[kostas@webtech-omen ~]$ 
```

Here's the gzipped version of the corrupted `GeoLite2-City.mmdb` (I just saved it before the update): https://pheix.org/pub/GeoLite2-City.mmdb.gz

**Comment by bbkr** [comment]
Confirmed bug.

**Tech note:** When reading file backwards to discover start of metadata I locate 0xAB first, then code should check if it begins magic `0xAB 0xCD 0xEF MaxMind.com` sequence. But for that wrong operator was used - checking temporary buffer length instead of buffer content. It always worked correctly by accident, because it is very rare case to get `0xAB` in encoded metadata control bytes. Until now - 2024-12-24 release of GeoLite2-City accidentaly contained first character of magic metadata start string.

Fix and release is on the way...

**Comment by bbkr** [comment]
Fixed and released as 1.1.4.

Thanks for the report and sorry for the inconvenience.
